kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java

```diff
@@ -278,4 +300,83 @@ private void setRecoveryNodeMetadataState(
             Instant.now().toEpochMilli());
     recoveryNodeMetadataStore.updateSync(updatedRecoveryNodeMetadata);
   }
+
+  /**
+   * Adjusts the offsets in the recovery task based on the availability of the offsets in Kafka.
+   * Returns <code>null</code> if the offsets specified in the recovery task are completely
+   * unavailable in Kafka.
+   */
+  private RecoveryTaskMetadata validateKafkaOffsets(RecoveryTaskMetadata recoveryTask) {
+    var topicPartition =
+        KaldbKafkaConsumer.getTopicPartition(
+            kaldbConfig.getKafkaConfig().getKafkaTopic(), recoveryTask.partitionId);
+    var earliestKafkaOffset = getPartitionOffset(topicPartition, OffsetSpec.earliest());
+    long newStartOffset = recoveryTask.startOffset;
+    long newEndOffset = recoveryTask.endOffset;
+
+    if (earliestKafkaOffset > recoveryTask.endOffset) {
+      LOG.warn(
+          "Entire task range ({}-{}) on topic {} is unavailable in Kafka (earliest offset: {})",
+          recoveryTask.startOffset,
+          recoveryTask.endOffset,
+          topicPartition,
+          earliestKafkaOffset);
+      return null;
+    }
+
+    var latestKafkaOffset = getPartitionOffset(topicPartition, OffsetSpec.latest());
+    if (latestKafkaOffset < recoveryTask.startOffset) {
+      // this should never happen, but if it somehow did, it would result in an infinite
+      // loop in the consumeMessagesBetweenOffsetsInParallel method
+      LOG.warn(
+          "Entire task range ({}-{}) on topic {} is unavailable in Kafka (latest offset: {})",
+          recoveryTask.startOffset,
+          recoveryTask.endOffset,
+          topicPartition,
+          latestKafkaOffset);
+      return null;
+    }
+
+    if (recoveryTask.startOffset < earliestKafkaOffset) {
+      LOG.warn(
+          "Partial loss of messages in recovery task. Start offset {}, earliest available offset {}",
+          recoveryTask.startOffset,
+          earliestKafkaOffset);
+      newStartOffset = earliestKafkaOffset;
+    }
+    if (recoveryTask.endOffset > latestKafkaOffset) {
+      // this should never happen, but if it somehow did, the requested recovery range should
+      // be adjusted down to the latest available offset in Kafka
+      LOG.warn(
+          "Partial loss of messages in recovery task. End offset {}, latest available offset {}",
+          recoveryTask.endOffset,
+          latestKafkaOffset);
+      newEndOffset = latestKafkaOffset;
+    }
+
+    return new RecoveryTaskMetadata(
```

Comment:
This seems a bit awkward - we technically just need the pair of updated offsets (or null) as we never directly use the returned `RecoveryTaskMetadata` object. This would be better if we had a dedicated object for this (ie, `KafkaOffsetPair`) instead of reusing the recovery task object.